src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,802 @@
+        match self.insert_builder.as_mut().unwrap().entry(ns) {
+            Entry::Occupied(mut entry) => entry.get_mut().append(document),
+            Entry::Vacant(entry) => {
+                let mut builder = InsertCommandBuilder::new(coll, self.max_entries);
+                builder.append(document);
+                entry.insert(builder);
+            }
+        }
+    }
+
+    fn upsert<'a>(&'a mut self, op: Op, row: RowRef<'a>) {
+        let mut document = self.document_from_row_ref(row);
+        let ns = self.extract_namespace_from_row_ref(row);
+        let coll = ns.1.clone();
+        let pk_index = self.pk_index.unwrap();
+        let pk_field = &self.schema.fields[pk_index];
+        let pk_datum = bson_from_scalar_ref(pk_field, row.datum_at(pk_index));
+        // Specify the primary key (_id) for the MongoDB collection if the user does not provide one.
+        if pk_field.name != "_id" {
+            document.insert("_id", pk_datum.clone());
+        }
+
+        match op {
+            Op::Insert | Op::UpdateInsert => {
+                match self.upsert_builder.as_mut().unwrap().entry(ns) {
+                    Entry::Occupied(mut entry) => entry.get_mut().append(pk_datum, document),
+                    Entry::Vacant(entry) => {
+                        let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                        builder.append(pk_datum, document);
+                        entry.insert(builder);
+                    }
+                }
+            }
+            Op::Delete | Op::UpdateDelete => {
+                match self.delete_builder.as_mut().unwrap().entry(ns) {
+                    Entry::Occupied(mut entry) => entry.get_mut().append(pk_datum),
+                    Entry::Vacant(entry) => {
+                        let mut builder = DeleteCommandBuilder::new(coll, self.max_entries);
+                        builder.append(pk_datum);
+                        entry.insert(builder);
+                    }
+                }
+            }
+        }
+    }
+
+    async fn write<'a>(&'a mut self, op: Op, row: RowRef<'a>) -> Result<()> {
+        if self.is_append_only {
+            if op != Op::Insert {
+                return Ok(());
+            }
+            self.append(row);
+        } else {
+            self.upsert(op, row);
+        }
+
+        self.buffered_entries += 1;
+        if self.buffered_entries >= self.max_entries {
+            self.flush().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn flush(&mut self) -> Result<()> {
+        if self.is_append_only {
+            if let Some(mut insert_builder) = self.insert_builder.take() {
+                for (ns, builder) in insert_builder.drain() {
+                    self.send_bulk_write_command(&ns.0, builder.build()).await?;
+                }
+                self.insert_builder = Some(insert_builder);
+            }
+        } else {
+            // 1. Writing deletes first
```

Comment:
Is there a problem here, such as changing from insert first and then delete to delete first and then insert, which prevents the data from being deleted

Reply:
Yes, we are doing batching here, and mongodb database command does not support mixing multiple upsert and delete operations in one go. So the bulk delete must be sent first to prevent data from being accidentally deleted by the adjacent `UpdateDelete` and `UpdateInsert`.

Reply:
if I want to 
insert  pk
delete pk
the result is no pk
but in there will be ->  delete pk insert pk 
the result is pk
so 
Should we need to use a map to cache the results here, perform column deletion, and then insert it

Reply:
You’re right, I haven’t been noticing this no-op case. But it seems we can simply ignore the UpdateDelete and then reverse the order, .i.e, sending bulk upsert first then the bulk delete, how do you think? 🤔

Reply:
It seems that there are still issues, such as` delete pk insert pk` -> pk
in there will be -> insert pk delete pk -> no pk

Reply:
understood, I will do some compaction logic for this case.

Reply:
`insert pk, delete pk` will be compact to a no-op before go out.
https://github.com/risingwavelabs/risingwave/pull/11070

Reply:
@xxhZs @StrikeW PTAL, the deletion building logic has now been merged in `UpsertCommandBuilder`. Under same pk, the pk that was previously marked for deletion will be reverted during the upsert. And the `UpdateDelete` is ignored in upsert mode as well.

Reply:
when the mongodb driver supports bulk-write API, we can make our code more elegant, because upsert and delete operations can be mixed in one go. As long as it looks like the [bulk-write](https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#Collection.BulkWrite) in go.